parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +607,342 @@ protected GroupType build(String name) {
+        if (grandParent != null) {
+          return (MapValueBuilder<P, Q>) new MapValueBuilder<P, Q>(parent, grandParent,
+              type).repetition(repetition);
+        } else {
+          return (MapValueBuilder<P, Q>) new MapValueBuilder<P, Q>(parent,
+              ((MapBuilder) parent).returnType, type).repetition(repetition);
+        }
+      }
+
+      public MapValueBuilder<P, Q> requiredValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.REQUIRED);
+      }
+
+      public MapValueBuilder<P, Q> optionalValue(PrimitiveTypeName type) {
+        return value(type, Type.Repetition.OPTIONAL);
+      }
+
+      public MapGroupValueBuilder<P, Q> value(Type.Repetition repetition) {
+        ((MapBuilder)parent).setKeyType(build("key"));
+        return (MapGroupValueBuilder<P, Q>) new MapGroupValueBuilder<P, Q>(parent,
+            grandParent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P, Q> requiredGroupValue() {
+        return value(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P, Q> optionalGroupValue() {
+        return value(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P, Q> extends BaseGroupBuilder<Q,
+        MapGroupValueBuilder<P, Q>> {
+      private final P parent;
+
+      public MapGroupValueBuilder(P parent, Q grandParent) {
+        super(grandParent);
+        this.parent = parent;
+      }
+
+      public Q named(String name) {
+        if (parent instanceof  MapBuilder) {
+          ((MapBuilder) parent).setValueType(build("value"));
+          return (Q) ((MapBuilder) parent).named(name);
+        }
+
+        throw new UnsupportedOperationException("named not called on correct TypeBuilder object");
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P, Q> self() {
+        return this;
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public MapBuilder(P parent) {
+      super(parent);
+    }
+
+    private MapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected MapBuilder<P> self() {
+      return this;
+    }
+
+    public MapKeyBuilder<MapBuilder<P>, P> key(PrimitiveTypeName type) {
+      return new MapKeyBuilder<MapBuilder<P>, P>(this, parent, type);
+    }
+
+    public MapGroupKeyBuilder<MapBuilder<P>, P> groupKey() {
+      return new MapGroupKeyBuilder<MapBuilder<P>, P>(this, parent);
+    }
+
+    @Override
+    protected Type build(String name) {
+      if (keyType == null) {
+        keyType = STRING_KEY;
+      }
+      return buildGroup(repetition)
+          .as(OriginalType.MAP)
+          .addField(repeatedGroup()
```

Comment:
calling `repeatedGroup()` on the group builder is identical to calling `addField` and passing in a type built from `repeatedGroup()`